### PR TITLE
initContainers: Investigating initContainers to be used in the operator

### PR DESCRIPTION
**Pull request by avisiedo**
Add a set of proof of concepts using initContainers to set up the environment before running the main container.

The details can be seen in the InitContainers.md document.

Some of them are not finished. The current in progress and pending paths will be handled in different branches to shrink the scope.

Any comment are welcomed.

---

- A makefile is provided to use the different PoC quick and easy.
- Some make rules need DOCKER_IMAGE to be defined to a valid image.
- make container-push need to be logged in the repository where the image will be pushed.
- To reference the poc, use the APP environment variable; `export APP=poc-05-a.yaml; make app-deploy get-info`.
- The traces could be seen by `oc logs pod/pod-id -c container-id`.

### CODE REVIEW

File: miscelanea/002-init-container/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,58 @@
+FROM docker.io/freeipa/freeipa-server:centos-8
+
+# Workaround for running systemd mode from podman
+RUN mkdir /usr/sbin/original \
+    && ln -svf /usr/lib/systemd/systemd /usr/sbin/original/init \
+    && rm /usr/sbin/init
+COPY init-wrapper.sh /usr/sbin/init
+COPY init-data /usr/loca/sbin/init
+
+# Workaround to don't loose the machine-id between different executions
```
Comment (avisiedo) [inline miscelanea/002-init-container/Dockerfile]: This is not needed. False positive when I got the report from dive tool. I will remove this block.

File: miscelanea/002-init-container/InitContainer.md [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+- Deployment of the application up to the point to run the main container/s
+  which only have to run systemd in the first stage, or independent containers
+  when we achieve to split all the different services. For achieving this it
+  will be used all the knowledge aquired with the previous point, and it will
+  be combined to create all the OpenShift objects needed to provide a
+  deployment just running "oc apply -f freeipa.yaml".
+- Shrink capabilities, this will need a deeper study and investigation as
+  systemd is involved on this. See the references at the end.
+
+As part of this investigation a final [freeipa.yaml](freeipa.yaml) object has
```
Comment (frasertweedale) [inline miscelanea/002-init-container/InitContainer.md]: `freeipa.yaml` is a dead link - the file does not exist in this directory.  Was it renamed or forgotten to commit?

File: miscelanea/002-init-container/InitContainer.md [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+
+- Private cloud which deploy Freeipa instance.
+- Public cloud which use the Freeipa instance form the Private cloud.
+  This will need to map an external service, and in this situation this
+  could be helpful.
+
+The proof of concept here just map an external gitlab instance to gather
+some information using a common internal service name.
+
+See: [pod-05-c.yaml](pod-05-c.yaml).
```
Comment (frasertweedale) [inline miscelanea/002-init-container/InitContainer.md]: Typo: s/pod/poc/.

File: miscelanea/002-init-container/InitContainer.md [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+
+### PENDING Redirecting traffic to the Pod using External IP
+
+TODO Fix the wrong behavior
+
+See: [poc-05-d.yaml](poc-05-d.yaml).
```
Comment (frasertweedale) [inline miscelanea/002-init-container/InitContainer.md]: Dead link - there is no `poc-05-d.yaml`.

File: miscelanea/002-init-container/InitContainer.md [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+
+## PENDING Application definition
+
+Use a StatefullSet object to make the deployment of the applications. For more
+informaiton about the reason, please look at "Chapter 11. Stateful Service" at
+"Kubernetes patterns" book.
+
+See: [freeipa.yaml](freeipa.yaml).
```
Comment (frasertweedale) [inline miscelanea/002-init-container/InitContainer.md]: Same as above, there is no `freeipa.yaml` in the commit.

File: miscelanea/002-init-container/InitContainer.md [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+
+### PENDING Redirecting traffic to the Pod using External IP
+
+TODO Fix the wrong behavior
+
+See: [poc-05-d.yaml](poc-05-d.yaml).
+
+## PENDING Application definition
+
+Use a StatefullSet object to make the deployment of the applications. For more
```
Comment (frasertweedale) [inline miscelanea/002-init-container/InitContainer.md]: Typo - s/Statefull/Stateful/

**Review by frasertweedale — COMMENTED** [review]
@avisiedo I finally did watch the recording and read this PR.  It is really great content, thank you so much!

There are a few nits; some typos and dead links.  Content is otherwise fine.  If we desire, we can make incremental improvements after the initial merge.

I request one addition.  We could mention that initContainers run every time the pod is (re)started, not only the first time the pod runs.  I was not sure what the behaviour would be and had to read the Kubernetes docs to find out.

File: miscelanea/002-init-container/InitContainer.md (hunk 2 above)
Comment (avisiedo) [reply miscelanea/002-init-container/InitContainer.md]: Making changes on this to remove the dead-link and adapt the content to the current state. Thanks.

**Comment by frasertweedale** [comment]
Thanks @avisiedo!  Merging.

**Pull request merged**